### PR TITLE
Terachem input file generator 

### PR DESCRIPTION
**Pull request by ffangliu**
@dgasmith Hi Daniel, the Terachem input file generator is added here. I created a pull request but later closed it because I found something to change in the test_compute.py. Sorry for the confusion.

Fang

### CODE REVIEW

**Comment by dgasmith** [comment]
It looks like you added another commit to the closed PR. You can continue to push new commits to your feature branch and they will show up on GitHub automatically.

**Comment by ffangliu** [comment]
Oh, I see. I used to think that pull request is specific to a revision, but actually it is for a branch.  This request and the closed one are actually the same branch.

**Comment by dgasmith** [comment]
It looks like you rebased then merged. @loriab do you remember a good location for a tutorial on fixing this?

**Comment by ffangliu** [comment]
> It looks like you rebased then merged. @loriab do you remember a good location for a tutorial on fixing this?
I first merged and then saw the discussions on rebasing for the molpro branch. Then I did rebase but then got conflicts and fixed it. Does the commit history still look incorrect? @dgasmith

**Comment by dgasmith** [comment]
My guess is that you hit conflicts when you went to `git push`, since the two histories are different a `git push --force` is required to overwrite your feature branches history. Currently you have an extra ~42 commits that should not be present. Probably the easiest thing to do is:
1) Make a new branch
2) `git cherry-pick hash` the new commits that you added over (like `14a57a0`)
3) `git pull --rebase upstream master`
4) `git push origin terachem`

Where origin is you fork and upstream is this repository.

**Comment by ffangliu** [comment]
Thanks for the suggestions. I'll try this out.

**Comment by ffangliu** [comment]
Sorry, I guess it didn't work. Since the parsing functions I'm adding are not closely related to the other parts, may I close this branch and start a brand new one?  Specifically, may I do the following steps:

1. delete my current forked repo
2. fork from the QCEngine/master
3. In the new forked repo, create the branch terachem
4. Add the parse functions there
5. create pull request from the new repo

Thanks,
Fang

**Comment by loriab** [comment]
https://github.com/psi4/psi4/pull/1490#issuecomment-457765390 is a set of instructions for escaping this git situation. No need to delete your fork or any branches, but you will be using a new branch locally (e.g., `new_terachem` where that link has `new_sinf`) and pushing to the branch associated with this PR (i.e., `terachem` where that link has `sinf`) The commits will be `66576b1` where the link has `b2b49f6`, then `8a78fd3`, ending with  `5ea3bef` for 6 total commits (the ones you authored).

**Comment by dgasmith** [comment]
This has been rebased, please do a `git reset --hard origin/terachem`.

**Comment by ffangliu** [comment]
@loriab Many thanks for the detailed instructions in your link.
@dgasmith I just noticed that you did the commit and fixed the problem. Thanks!

**Comment by dgasmith** [comment]
It might be worth looking into #29 to see future directions for this PR.

**Comment by ffangliu** [comment]
Just added the output parser.

**Comment by dgasmith** [comment]
Can you integrate in the executor and testing technology as shown in #29 which would correspond to the QCEngineRecords (Vault) repository?

**Comment by ffangliu** [comment]
Sure. I'll try that out

**Comment by ffangliu** [comment]
I have applied the executor. For the output parser part, I'm not sure how the output_data dictionary should be formatted, so I just temporarily assigned the values with the keys that present their physical meaning. Those need to be later changed to match the format of QCSchema.

**Comment by dgasmith** [comment]
You can see a list of currently supported properties [here](https://molssi-qc-schema.readthedocs.io/en/latest/auto_props.html). Please also fix the LGTM alerts when you have a chance, you can see the errors by click on `view on LGTM.com`.

**Comment by ffangliu** [comment]
Fixed the output_data format to match QCSchema, and all LGTM alerts are now removed.

**Comment by ffangliu** [comment]
Here are the environment variables needed for terachem execution:

```
# Set TeraChem env variable. Needed for TeraChem to find the basis sets, etc
TeraChem="your_terachem_installation_path"

# Set OpenMM env variable. 
OpenMM="your_openMM_installation_root"
OPENMM_INCLUDE_PATH=$OpenMM/include
OPENMM_LIB_PATH=$OpenMM/lib
OPENMM_PLUGIN_DIR=$OpenMM/lib/plugins

# Set env for the NBO executable. TeraChem population analysis calculation is done by driving the third party NBO executable distributed with TeraChem.
NBOEXE="your_path_to_nbo/nbo6.i4.exe"

# Set cuda related 
CUDA_HOME=/opt/cuda8.0
CUDA_DIR=/opt/cuda8.0

# Set Intel related 
IHOME=/opt/intel/2016/compilers_and_libraries/linux
MKLROOT=$IHOME/mkl
TBBROOT=$IHOME/tbb
INTEL_LICENSE_FILE="your-path-to-license-file"

# Set mpich2 related
MPI_HOME=/opt/mpich2/1.4.1p1
MPI_BIN=$MPI_HOME/bin
MPI_FORTRAN_MOD_DIR=y$MPI_HOME/include
MPI_INCLUDE=$MPI_HOME/include
 MPI_LIB=$MPI_HOME/lib
MPI_PYTHON_SITEARCH=$MPI_HOME/bin
MPI_SUFFIX=_mpich2


# Prepend include:
INCLUDE=$MKLROOTDIR/include:$INCLUDE
INCLUDE=$CUDA_HOME/include:$INCLUDE

# Prepend dynamic libraries
LD_LIBRARY_PATH=$TeraChem/lib:$LD_LIBRARY_PATH 
LD_LIBRARY_PATH =$OpenMM/lib:$LD_LIBRARY_PATH 
LD_LIBRARY_PATH=$CUDA_HOME/lib64:$LD_LIBRARY_PATH 
LD_LIBRARY_PATH=$IHOME/lib/intel64:$LD_LIBRARY_PATH
LD_LIBRARY_PATH=$MKLROOT/lib/intel64:$LD_LIBRARY_PATH
LD_LIBRARY_PATH=$IHOME/tbb/lib/intel64/gcc4.4:$LD_LIBRARY_PATH
LD_LIBRARY_PATH=$IHOME/ipp/lib/intel64:$LD_LIBRARY_PATH
LD_LIBRARY_PATH $IHOME/ipp/../compiler/lib/intel64:$LD_LIBRARY_PATH


# Prepend static libraries
LIBRARY_PATH=$OpenMM/lib:$LIBRARY_PATH
LIBRARY_PATH=$IHOME/lib/intel64:$LIBRARY_PATH
LIBRARY_PATH=$MKLROOT/lib/intel64:$LIBRARY_PATH
LIBRARY_PATH=$IHOME/tbb/lib/intel64/gcc4.4:$LIBRARY_PATH
LIBRARY_PATH=$IHOME/ipp/lib/intel64:$LIBRARY_PATH
LIBRARY_PATH=$IHOME/ipp/../compiler/lib/intel64:$LIBRARY_PATH
LD_LIBRARY_PATH=$IHOME/mpi/lib64:$LIBRARY_PATH


# Prepend TeraChem executable path to PATH
PATH=$TeraChem/bin:$PATH
PATH=$OpenMM/bin:$PATH
PATH=$CUDA_HOME/bin:$PATH
PATH =$IHOME/bin/intel64:$PATH
PATH=$IHOME/mpi/bin64:$PATH
PATH =$MPI_HOME/bin

# Prepend CPATH
CPATH=$OpenMM/include:$CPATH
 CPATH=$MKLROOTDIR/include:$CPATH
CPATH=$IHOME/tbb/include:$CPATH
CPATH=$IHOME/ipp/include:$CPATH

```

File: qcengine/programs/tests/test_terachem.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+     inp = qcel.models.ResultInput.parse_raw(data["input.json"])
+
+     output = qcng.get_program('terachem').parse_output(data, inp)
+
+     output_ref = qcel.models.Result.parse_raw(data["output.json"])
+
+     assert compare_recursive(output_ref.dict(), output.dict()) 
+
+@pytest.mark.parametrize('test_case', terachem_info.list_test_cases())
+def test_molpro_input_formatter(test_case):
```
Comment (dgasmith) [inline qcengine/programs/tests/test_terachem.py]: `test_molpro_input_formatter` -> `test_terachem_input_formatter`

File: qcengine/programs/terachem.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    def compute(self, input_data: 'ResultInput', config: 'JobConfig') -> 'Result':
+        if self.found() == False:
+            raise ImportError("Could not find terachem in the envvar path.")
+        input_data = input_data.copy().dict()
+        # Setup the job
+        job_inputs = self.build_input(input_data, config)
+        # Run terachem
+        exe_outputs = execute(job_inputs)
+        [exe_success, proc] = exe_outputs
```
Comment (dgasmith) [inline qcengine/programs/terachem.py]: No need for `[...]` syntax, `...` is fine.

**Comment by ffangliu** [comment]
Sorry I didn't see your comments when pushing these changes. Now I'm fixing the things you commented and testing. Will let you know.

**Comment by ffangliu** [comment]
Should be ready for merge this time (we can wait for LGTM).  I tested on our cluster. The compute module can return a FailedOperation instance, with messages about terachem not found since environment variables are not set.

**Comment by dgasmith** [comment]
Looks great! Thanks for the addition. We will need to do more rigorous testing soon, but a good first draft.

**Pull request merged**